pkg/destinations/destinations.go

```diff
@@ -1,28 +1,66 @@
 package destinations
 
 import (
+	"fmt"
 	"io"
 	"scratchdata/models"
 	"scratchdata/pkg/destinations/clickhouse"
 	"scratchdata/pkg/destinations/duckdb"
 	"scratchdata/pkg/destinations/memory"
-	"scratchdata/util"
+	"sync"
 )
 
-func GetDestination(dbConfig models.DatabaseConnection) DatabaseServer {
-	configType := dbConfig.Type
-	connectionSettings := dbConfig.ConnectionSettings
+var (
+	destinationCache = struct {
+		sync.RWMutex
+		m map[string]DatabaseServer
+	}{
+		m: map[string]DatabaseServer{},
+	}
+)
+
+func deriveDatabaseConnectionKey(dbConfig models.DatabaseConnection) string {
+	// if ID isn't enough, we can also marshal dbConfig using a canonical/deterministic JSON encoder
+	return dbConfig.ID
+}
 
-	switch configType {
+func openDBServer(dbType string, settings map[string]any) (DatabaseServer, error) {
+	switch dbType {
 	case "duckdb":
-		return util.ConfigToStruct[duckdb.DuckDBServer](connectionSettings)
+		return duckdb.OpenServer(settings)
 	case "clickhouse":
-		return util.ConfigToStruct[clickhouse.ClickhouseServer](connectionSettings)
+		return clickhouse.OpenServer(settings)
 	case "memory":
-		return util.ConfigToStruct[memory.MemoryDBServer](connectionSettings)
+		return memory.OpenServer(settings), nil
 	default:
-		return nil
+		return nil, fmt.Errorf("GetDestination: Unsupported database type: %s", dbType)
+	}
+}
+
+func GetDestination(dbConfig models.DatabaseConnection) (DatabaseServer, error) {
+	key := deriveDatabaseConnectionKey(dbConfig)
+
+	destinationCache.RLock()
```

Comment:
would `sync.Map` be useful here?

Reply:
It's not ideal in this case because we want to block the caller, so that if N calls are made to the same ID, we only make 1 connection. Thinking about it more, I _think_ the number of connection configs will be small, so I removed the RWMutex and just use Mutex since after the first connection succeeds we'll only being doing a map lookup. If performance can become a problem it's easy to implement a slightly more complex method that blocks based on the key only.